src/NKSolver/NKSolvers.F90

```diff
@@ -589,7 +589,17 @@ subroutine NKStep(firstCall)
     ! convergnce check can't do anything either. By multiplying by
     ! 0.5, we make sure that the linear solver actually has to do
```

Comment:
Let me clarify; The linear residual is of similar order as the nonlinear residual because the linear system's RHS is actually equal to the nonlinear system's residuals. Even though they are not identical (the linear system may have an initial guess, the way you compute the norms can be different, etc etc) they have a similar magnitude.

In here, the term "oversolving" refers to a different kind of oversolving prevented by the EW. The EW algorithm makes sure that you are not solving the linear system too tightly when you are far from the solution. The idea is that you are better of taking a more approximate step rather than taking a very accurate step, which will not give you too much nonlinear convergence to begin with. Here, what we mean by oversolving is this: Assume you are at a relative nonlinear convergence of 1e-11, and your target nonlinear convergence is 1e-12. You go into your Newton solver and now you need to solve a linear system. EW algorithm knows that you are close to the solution, so it will give you a relative convergence of a very low value, say 1e-6. Now, because you were so close to the L2 convergence target, you most likely could just solve the linear system by 0.1, and that will probably give you the nonlinear convergence you need. So this is what we mean by avoiding the oversolve. If you dont have an atol here, you will need to solve the linear system to 1e-6, whereas you could just do a bit of work and quit early because you are already there in the L2 convergence.

Now, there is another pitfall here. Lets go back to the previous numbers, you are at 1e-11 L2 convergence and target is 1e-12. If you set the atol value as `totalR0*L2Conv`, then the linear solver will do about 0.1 convergence and then quit. But nonlinear convergence can only be as good as your linear convergence, and often it is slightly worse than your linear convergence because the linear convergence is a linear approximation of your nonlinear system. So if you converge your linear system to 0.1, your nonlinear system may converge by 0.11. Now think what the solver would do; The top level nonlinear solver thinks there is still progress to be done since you are not exactly at 1e-12 (you were at 1e-11 and only converged 0.11 more), and the linear sytem thinks you are below the absolute tolerance for linear convergence because of you you just defined this atol. So they just go back and forth until you stop the code (it will run out of total iterations and quit with a "failed" solve). To avoid this issue, we multiply the atol by 0.01, so that we actually guarantee to solve past the nonlinear target, but only go past this target by 2 orders of magnitude (instead like 6 or 8). 

FINALLY, the complex system changes things a bit. Remember, we are trying to avoid oversolving the linear system, and we determine this by the residual. When running in complex mode, the complex residuals will lag, and we are interested in maybe doing more iterations to the solver to converge the complex residual, even though the real residual is already converged. However, if we keep the same atol of 0.01 lower than the L2 target, we can only solve the linear system by 0.01! Instead, I would much rather doing a tight solve (I can manually set the newton linear convergence to 1e-6) and nail down the complex residual, so that is why I have a different atol in the complex mode. This problem is caused by the complex residuals being negligible compared to real residuals, even when the real residuals are at machine zero.

This turned into a lecture, but whatever. Does this help @sseraj ? I wont spell check this so let me know if there is stuff thats not clear still. I would appreciate if you can suggest some changes to the comment to clarify this for people new to these ideas, or maybe we can link this comment in the code?

Reply:
Thanks for writing this up! This helps a lot. I would suggest 3 changes to the comments to clear things up:
1) Add that the linear and nonlinear residuals have similar magnitude _at the start of the linear solution_, assuming a small initial guess
2) Mention that KSPSolve exits when either the rtol or atol is satisfied (not both), which means that the atol only comes into play near the end of the nonlinear solution
3) The block comment in NKStep does not actually mention oversolving the system so I would change your comment to just say "in the real mode, we set the atol slightly lower than the target L2 convergence as detailed above"

Reply:
Done. Please take a look at the new version

Reply:
The new comments look great. Thanks! There are some typos left over from before but nothing major.